cirq-core/cirq/ops/boolean_hamiltonian.py

```diff
@@ -112,6 +114,171 @@ def _decompose_(self):
         )
 
 
+def _gray_code_comparator(k1: Tuple[int, ...], k2: Tuple[int, ...], flip: bool = False) -> int:
+    """Compares two Gray-encoded binary numbers.
+
+    Args:
+        k1: A tuple of ints, representing the bits that are one. For example, 6 would be (1, 2).
+        k2: The second number, represented similarly as k1.
+        flip: Whether to flip the comparison.
+
+    Returns:
+        -1 if k1 < k2 (or +1 if flip is true)
+        0 if k1 == k2
+        +1 if k1 > k2 (or -1 if flip is true)
+    """
+    max_1 = k1[-1] if k1 else -1
+    max_2 = k2[-1] if k2 else -1
+    if max_1 != max_2:
+        return -1 if (max_1 < max_2) ^ flip else 1
+    if max_1 == -1:
+        return 0
+    return _gray_code_comparator(k1[0:-1], k2[0:-1], not flip)
+
+
+def _simplify_commuting_cnots(
+    cnots: List[Tuple[int, int]], flip_control_and_target: bool
+) -> Tuple[bool, List[Tuple[int, int]]]:
+    """Attempts to commute CNOTs and remove cancelling pairs.
+
+    Commutation relations are based on 9 (flip_control_and_target=False) or 10
+    (flip_control_target=True) of [4]:
+    When flip_control_target=True:
+
+         CNOT(j, i) @ CNOT(j, k) = CNOT(j, k) @ CNOT(j, i)
+    ───X───────       ───────X───
+       │                     │
+    ───@───@───   =   ───@───@───
+           │             │
+    ───────X───       ───X───────
+
+    When flip_control_target=False:
+
+    CNOT(i, j) @ CNOT(k, j) = CNOT(k, j) @ CNOT(i, j)
+    ───@───────       ───────@───
+       │                     │
+    ───X───X───   =   ───X───X───
+           │             │
+    ───────@───       ───@───────
+
+    Args:
+        cnots: A list of CNOTS, encoded as integer tuples (control, target).
+        flip_control_and_target: Whether to flip control and target.
+
+    Returns:
+        A Boolean that tells whether a simplification has been performed.
+        The CNOT list, potentially simplified.
```

Comment:
Please clarify that this returns a tuple containing these two elements.

Reply:
Like this?

Reply:
I should have been clearer - I meant something like:
```
Returns:
    A tuple containing a boolean {...} and the CNOT list {...}.
```

Reply:
Done.